lib/proxy/client.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	if c.TLSConfig == nil {
+		return trace.BadParameter("missing tls config")
+	}
+
+	if len(c.TLSConfig.Certificates) == 0 {
+		return trace.BadParameter("missing tls certificate")
+	}
+
+	if c.getConfigForServer == nil {
+		c.getConfigForServer = getConfigForServer(c.TLSConfig, c.AccessPoint, c.Log)
+	}
+
+	return nil
+}
+
+// clientConn hold info about a dialed grpc connection
+type clientConn struct {
+	*grpc.ClientConn
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+
+	id   string
+	addr string
+}
+
+// Client is a peer proxy service client using grpc and tls.
+type Client struct {
+	sync.RWMutex
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config  ClientConfig
+	conns   map[string]*clientConn
+	metrics *clientMetrics
+}
+
+type connections struct{}
+
+// NewClient creats a new peer proxy client.
+func NewClient(config ClientConfig) (*Client, error) {
+	err := config.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	metrics, err := newClientMetrics()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	closeContext, cancel := context.WithCancel(config.Context)
+
+	c := &Client{
+		config:  config,
+		ctx:     closeContext,
+		cancel:  cancel,
+		conns:   make(map[string]*clientConn),
+		metrics: metrics,
+	}
+
+	if c.config.sync != nil {
+		go c.config.sync()
+	} else {
+		go c.sync()
+	}
+
+	return c, nil
+}
+
+// sync runs the proxy watcher and cleanup functionalities.
+func (c *Client) sync() {
+	ticker := c.config.Clock.NewTicker(defaults.ResyncInterval)
+	defer ticker.Stop()
+
+	proxyWatcher, err := services.NewProxyWatcher(c.ctx, services.ProxyWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.Component(teleport.ComponentProxyPeer),
+			Client:    c.config.AccessPoint,
+			Log:       c.config.Log,
+		},
+	})
+	if err != nil {
+		c.config.Log.Errorf("Error initializing proxy peer watcher: %+v.", err)
+		return
+	}
+	defer proxyWatcher.Close()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			c.config.Log.Debug("Stopping peer proxy sync.")
+			return
+		case proxies := <-proxyWatcher.ProxiesC:
+			if err := c.updateConnections(proxies); err != nil {
+				c.config.Log.Errorf("Error syncing peer proxies: %+v.", err)
+			}
+		case <-ticker.Chan():
+			if err := c.cleanup(); err != nil {
+				c.config.Log.Errorf("Error cleaning up peer proxies: %+v.", err)
+			}
+		}
+	}
+}
+
+func (c *Client) updateConnections(proxies []types.Server) error {
+	c.Lock()
+	defer c.Unlock()
+	var errs []error
+
+	toDial := make(map[string]types.Server)
+	for _, proxy := range proxies {
+		toDial[proxy.GetName()] = proxy
+	}
+
+	for id, conn := range c.conns {
+		_, ok := toDial[id]
+
+		// delete nonexistent connections
+		if !ok {
+			c.closeConn(conn)
+			continue
+		}
+
+		// test existing connections
+		if err := c.testConnection(conn); err != nil {
+			errs = append(errs, err)
+			c.closeConn(conn)
+			continue
+		}
```

Comment:
I don't think we need to test the connection. The `grpc.ClientConn` should reconnect automatically if there is a transient failure. However we may want to check if the `PeerAddr` has changed and close/reconnect in that case.

Reply:
Oh ok, I assumed that for the proxy peer addr to change the proxy needs to be restarted, and assumed it gets a new id after a restart so it wouldn't be necessary to check for that case. 

as for the connection testing, the connection state isn't always up to date and we need to do something with it for GetState to return the correct one, that's why testConnection is there. Maybe I should check the error returned or the state right after and keep the connection in case it's a transient failure and not something more permanent.

Reply:
For cloud the peer address changing probably isn't an issue but could be in other environments.